crashing dump for hparams containing fsspec

## 🐛 Bug

If you pass a model as PL model argument and dum this model, it crashes...

## Please reproduce using [the BoringModel and post here]

Failing several models in Bolts
https://github.com/PyTorchLightning/pytorch-lightning-bolts/runs/1254957648
sample of failing test:
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/ccdc9f9952153abf9b12f00e05294fa332d5f424/tests/models/test_vision.py#L8-L31

### Additional context

```
        cls = type(data)
        if cls in copyreg.dispatch_table:
            reduce = copyreg.dispatch_table[cls](data)
        elif hasattr(data, '__reduce_ex__'):
>           reduce = data.__reduce_ex__(2)
E           TypeError: __reduce_ex__() takes exactly one argument (0 given)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed `hparams` saving - save the state when `save_hyperparameters()` is called [in `__init__`] ([#4163](https://github.com/PyTorchLightning/pytorch-lightning/pull/4163))
 
+- Fixed runtime failure while exporting `hparams` to yaml ([#4158](https://github.com/PyTorchLightning/pytorch-lightning/pull/4158))
 
 
 ## [1.0.1] - 2020-10-14

--- a/pytorch_lightning/core/saving.py
+++ b/pytorch_lightning/core/saving.py
@@ -381,6 +381,7 @@ def save_hparams_to_yaml(config_yaml, hparams: Union[dict, Namespace]) -> None:
             yaml.dump(v)
         except TypeError as err:
             warn(f"Skipping '{k}' parameter because it is not possible to safely dump to YAML.")
+            hparams[k] = type(v).__name__
         else:
             hparams_allowed[k] = v
 

--- a/pytorch_lightning/core/saving.py
+++ b/pytorch_lightning/core/saving.py
@@ -18,6 +18,7 @@ import inspect
 import os
 from argparse import Namespace
 from typing import Union, Dict, Any, Optional, Callable, MutableMapping
+from warnings import warn
 
 import fsspec
 import torch
@@ -372,10 +373,20 @@ def save_hparams_to_yaml(config_yaml, hparams: Union[dict, Namespace]) -> None:
                     OmegaConf.save(OmegaConf.create(hparams), fp, resolve=True)
                 return
 
-    # saving the standard way
     assert isinstance(hparams, dict)
+    hparams_allowed = {}
+    # drop paramaters which contain some strange datatypes as fsspec
+    for k, v in hparams.items():
+        try:
+            yaml.dump(v)
+        except TypeError as err:
+            warn(f"Skipping '{k}' parameter because it is not possible to safely dump to YAML.")
+        else:
+            hparams[k] = v
+
+    # saving the standard way
     with fs.open(config_yaml, "w", newline="") as fp:
-        yaml.dump(hparams, fp)
+        yaml.dump(hparams_allowed, fp)
 
 
 def convert(val: str) -> Union[int, float, bool, str]:

--- a/pytorch_lightning/core/saving.py
+++ b/pytorch_lightning/core/saving.py
@@ -382,7 +382,7 @@ def save_hparams_to_yaml(config_yaml, hparams: Union[dict, Namespace]) -> None:
         except TypeError as err:
             warn(f"Skipping '{k}' parameter because it is not possible to safely dump to YAML.")
         else:
-            hparams[k] = v
+            hparams_allowed[k] = v
 
     # saving the standard way
     with fs.open(config_yaml, "w", newline="") as fp:

--- a/tests/models/test_hparams.py
+++ b/tests/models/test_hparams.py
@@ -18,6 +18,7 @@ from argparse import Namespace
 import cloudpickle
 import pytest
 import torch
+from fsspec.implementations.local import LocalFileSystem
 from omegaconf import OmegaConf, Container
 from torch.nn import functional as F
 from torch.utils.data import DataLoader
@@ -579,3 +580,22 @@ def test_init_arg_with_runtime_change(tmpdir):
     path_yaml = os.path.join(trainer.logger.log_dir, trainer.logger.NAME_HPARAMS_FILE)
     hparams = load_hparams_from_yaml(path_yaml)
     assert hparams.get('running_arg') == 123
+
+
+class UnsafeParamModel(BoringModel):
+    def __init__(self, my_path, any_param=123):
+        super().__init__()
+        self.save_hyperparameters()
+
+
+def test_model_with_fsspec_as_parameter(tmpdir):
+    model = UnsafeParamModel(LocalFileSystem(tmpdir))
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=2,
+        limit_val_batches=2,
+        limit_test_batches=2,
+        max_epochs=1,
+    )
+    trainer.fit(model)
+    trainer.test()

